### ISSUE TITLE
Xeon Phi round 2

### ISSUE THREAD

**Issue by raver119** [deeplearning4j/libnd4j#302]
TL/DR: many slow cores, with different clustering setups available. This investigation was taken over Quadrant clustering mode and Flat HBM memory mode.

To test performance with KNL we’ve used multiple synthetic tests available at backends-comparison repo. Results are available at Skymind benchmarking system.
Additionally, few more simple tests were added:
Java: https://github.com/deeplearning4j/backends-comparison/blob/e53cd24883cb19699f4f99f0c91380a17b3db9ec/benchmarks-pack/src/main/java/org/nd4j/benchmarks/toolbox/xeon/ShortScalar.java#L15-L15
C: https://gist.github.com/raver119/20d9adc1e5c9abcf4e04184000a1ac59

Compile args for ICC: icc sanity_check2.c -O2 -fp-model fast -qopenmp -Wall -lrt -xCOMMON-AVX512 -std=c99
Compile args for GCC: gcc sanity_check2.c -lrt -fopenmp -march=native -O2 --fast-math

Both tests are imitating scalarMul operations: B[i] = A[i] \* scalar; applied to 64k-elements array.

**Results**

_**HBM:**_
Gives 10% performance boost in average with current configuration.

_**i7 vs KNL:**_
In current KNL setup, it’s ±3 times slower then average 4 core i7 cpu.

Cache hit vs miss:
In both scenarios performance gap stays the same.

Cache hit op:
     i7: ~10 us
     KNL: ~40 us

Additional note here: for cache hit scenario, performance seems to be slightly better without omp threads.

Cache miss op:
     i7: ~35 us
     KNL: ~120 us

_**Cores vs Performance:**_
In most of cases best performance is achieved with clustering boundaries, scaling across clustering boundaries isn’t linear. So best performance for op usually achieved around 8-16 cores .
Here’s results that confirm this evidence: https://gist.github.com/raver119/a823b4104156905f081231d16365ab13
Pretty same results were achieved using C test code.

_**Memory bound vs computation bound:**_
To test comp.bound setup, expf() and cosf() was used it the same loop. Results there are pretty limited to scalarMul, however, slightly (around 7%) better results were achieved with 12 threads. That gives an idea of actual hardware limitations behind the scene.

**Conclusions**

Without explicit control over NUMA, KNL seems to be underperforming with our usual workloads/data dimensions. Probably this device still might be used with data parallelism or with special inference setups.

**Next Steps**

After discussing current situation with Paul (aka @treo), we’ve decided that we need to test strict NUMA-aware approach, with SNC-4 clustering mode and HBM-as-cache mode enabled. So let’s ask Intel to reboot that server and enable SNC-4 mode in BIOS.

**Comment by raver119** [comment]
This investigation follows original one, issued by @treo : https://github.com/deeplearning4j/libnd4j/issues/281

**Comment by treo** [comment]
A note on using svml: When using the array from https://gist.github.com/treo/3fd1390da7c0f33dd1b023409bf97cf1

Transforms.pow(arr, 2) results in a lot of NaNs.
